File: src/lib/githubUrlVerifier.js
/**
 * Verifies a URL or path and extracts owner/repo or parent/name information based on the source type.
 *
 * @param {string} urlString The URL or path string to verify and parse.
 * @param {0 | 1 | 2} sourceType - 0 for GitHub, 1 for GitLab, 2 for Local File.
 * @returns {{owner: string, repo: string}} An object containing the owner/group/parent and repo/project/name.
 * @throws {Error} If the URL/path is invalid for the given source type.
 */
export function verifyAndExtractSourceInfo(urlString, sourceType) {
    if (typeof urlString !== 'string' || urlString.trim() === '') {
        throw new Error('Input string cannot be empty.');
    }

    switch (sourceType) {
        case 0: // GitHub
            try {
                const url = new URL(urlString);
                const host = url.hostname;
                if (host !== 'github.com' && host !== 'www.github.com') {
                    throw new Error('URL is not from GitHub (github.com or www.github.com)');
                }

                const pathSegments = url.pathname.split('/').filter(segment => segment.length > 0);
                if (pathSegments.length < 2) {
                    throw new Error('GitHub URL must contain at least an owner and a repository name.');
                }

                const owner = pathSegments[0];
                let repo = pathSegments[1];
                if (repo.endsWith('.git')) {
                    repo = repo.slice(0, -4);
                }

                const ownerRegex = /^[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?$/;
                const repoRegex = /^[a-zA-Z0-9_.-]+$/; // GitHub repos can have dots
                const noDoubleHyphen = /--/;

                if (!ownerRegex.test(owner) || noDoubleHyphen.test(owner)) {
                    throw new Error(`Invalid GitHub owner name: "${owner}". Must be alphanumeric or single hyphens, not starting/ending with a hyphen, and no consecutive hyphens.`);
                }
                if (!repoRegex.test(repo) || repo.startsWith('-') || repo.endsWith('-') || repo.endsWith('.') || noDoubleHyphen.test(repo)) {
                    throw new Error(`Invalid GitHub repository name: "${repo}". Contains invalid characters or patterns.`);
                }
                return { owner, repo };
            } catch (e) {
                if (e instanceof TypeError && e.message.includes('Invalid URL')) { // Catch URL constructor errors
                     throw new Error(`Invalid GitHub URL format: ${urlString}`);
                }
                throw e; // Re-throw other errors (like custom ones)
            }

        case 1: // GitLab
            try {
                const url = new URL(urlString);
                const host = url.hostname;
                if (host !== 'gitlab.com' && host !== 'www.gitlab.com') {
                    throw new Error('URL is not from GitLab (gitlab.com or www.gitlab.com)');
                }

                const pathSegments = url.pathname.split('/').filter(segment => segment.length > 0);
                
                if (pathSegments.length === 0) {
                    throw new Error('GitLab URL path is empty.');
                }

                let repoWithGit = pathSegments.pop();
                if (typeof repoWithGit === 'undefined') { // Should not happen if length > 0 check passed, but defensive
                    throw new Error('Could not determine GitLab project name.');
                }
                let repo = repoWithGit.endsWith('.git') ? repoWithGit.slice(0, -4) : repoWithGit;

                if (pathSegments.length === 0) {
                    throw new Error('GitLab URL must contain at least a group/owner and a project name.');
                }
                const owner = pathSegments.join('/');

                // GitLab names: letters, digits, '_', '.', '-', '+'. No leading/trailing hyphens.
                const gitlabNamePartRegex = /^[a-zA-Z0-9](?:[a-zA-Z0-9_.\-+]*[a-zA-Z0-9])?$/; // Corrected regex
                const noDoubleHyphen = /--/;

                owner.split('/').forEach(segment => {
                    if (!gitlabNamePartRegex.test(segment) || noDoubleHyphen.test(segment)) {
                        throw new Error(`Invalid GitLab group/owner segment: "${segment}".`);
                    }
                });
                if (!gitlabNamePartRegex.test(repo) || repo.length === 0 || repo.endsWith('.') || noDoubleHyphen.test(repo)) {
                    throw new Error(`Invalid GitLab project name: "${repo}".`);
                }
                return { owner, repo };
            } catch (e) {
                 if (e instanceof TypeError && e.message.includes('Invalid URL')) {
                     throw new Error(`Invalid GitLab URL format: ${urlString}`);
                }
                throw e;
            }

        case 2: // Local File
            // Basic validation for local paths: just ensure it's not obviously empty or just whitespace.
            // More complex validation (e.g., existence) is out of scope for this client-side function.
            // For local paths, we'll treat the last segment as 'repo' and the preceding path as 'owner'.
            const normalizedPath = urlString.replace(/\\\\/g, '/'); // Normalize backslashes to forward slashes
            const segments = normalizedPath.split('/').filter(s => s.length > 0);

            if (segments.length === 0) {
                // Handle root paths like "/" or "C:/"
                if (normalizedPath === '/' || /^[a-zA-Z]:\/$/.test(normalizedPath)) {
                    return { owner: '', repo: normalizedPath };
                }
                throw new Error('Local path is empty or invalid.');
            }

            const repo = segments.pop();
            const owner = segments.join('/');
            
            if (!repo) { // Should not happen if segments.length > 0 initially
                 throw new Error('Could not determine file/directory name from local path.');
            }

            return { owner: owner || '.', repo }; // Use '.' for owner if no parent path after splitting

        default:
            throw new Error('Invalid source type specified. Use 0 for GitHub, 1 for GitLab, 2 for Local File.');
    }
}

File: src-tauri/src/github_url_verifier.rs
use serde::Serialize;
use regex::Regex;
use std::path::Path; // Added for local path manipulation

#[derive(Serialize, Debug)]
pub struct SourceInfo { // Renamed from GithubInfo
    pub owner: String, // For git hosts: owner/group. For local: parent directory.
    pub repo: String,  // For git hosts: repo name. For local: file/directory name.
}

#[tauri::command]
// Renamed function and added source_type parameter
pub fn verify_and_extract_source_info(url_str: &str, source_type: i32) -> Result<SourceInfo, String> {
    match source_type {
        0 => { // GitHub
            let url = url::Url::parse(url_str).map_err(|e| format!("Invalid GitHub URL: {}", e))?;
            let host = url.host_str().unwrap_or_default();
            if host != "github.com" && host != "www.github.com" {
                return Err("URL is not from GitHub (github.com or www.github.com)".to_string());
            }

            let mut path_segments: Vec<&str> = url
                .path_segments()
                .map_or_else(Vec::new, |segments| segments.filter(|s| !s.is_empty()).collect());

            if path_segments.len() < 2 {
                return Err("GitHub URL must contain at least an owner and a repository name.".to_string());
            }

            let owner = path_segments.remove(0).to_string();
            let repo = path_segments.remove(0).to_string().trim_end_matches(".git").to_string();


            // GitHub username: alphanumeric or single hyphens, not at start/end.
            // GitHub repo: alphanumeric, ., -, _
            let owner_regex = Regex::new(r"^[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?$").unwrap();
            let repo_regex = Regex::new(r"^[a-zA-Z0-9_.-]+$").unwrap();
            let no_double_hyphen = Regex::new(r"--").unwrap();

            if !owner_regex.is_match(&owner) || no_double_hyphen.is_match(&owner) {
                return Err(format!("Invalid GitHub owner name: '{}'. Must be alphanumeric or single hyphens, not starting/ending with a hyphen, and no consecutive hyphens.", owner));
            }
            if !repo_regex.is_match(&repo) || repo.starts_with('-') || repo.ends_with('-') || repo.ends_with('.') || no_double_hyphen.is_match(&repo) {
                 return Err(format!("Invalid GitHub repository name: '{}'. Contains invalid characters or patterns.", repo));
            }
            Ok(SourceInfo { owner, repo })
        }
        1 => { // GitLab
            let url = url::Url::parse(url_str).map_err(|e| format!("Invalid GitLab URL: {}", e))?;
            let host = url.host_str().unwrap_or_default();
            if host != "gitlab.com" && host != "www.gitlab.com" {
                return Err("URL is not from GitLab (gitlab.com or www.gitlab.com)".to_string());
            }

            let mut path_segments: Vec<String> = url
                .path_segments()
                .map_or_else(Vec::new, |segments| segments.filter(|s| !s.is_empty()).map(String::from).collect());

            if path_segments.is_empty() {
                 return Err("GitLab URL path is empty.".to_string());
            }
            
            let repo_name_with_git = path_segments.pop().unwrap_or_default();
            let repo = repo_name_with_git.trim_end_matches(".git").to_string();

            if path_segments.is_empty() {
                return Err("GitLab URL must contain at least a group/owner and a project name.".to_string());
            }
            let owner = path_segments.join("/");

            // GitLab names: letters, digits, emojis (not handled by simple regex), '_', '.', '-', '+'.
            // Cannot start with '-' or end with '.', '.git', '.atom'.
            let gitlab_name_part_regex = Regex::new(r"^[a-zA-Z0-9](?:[a-zA-Z0-9_.-+]*[a-zA-Z0-9])?$").unwrap();
            let no_double_hyphen = Regex::new(r"--").unwrap();

            for segment in owner.split('/') {
                if !gitlab_name_part_regex.is_match(segment) || no_double_hyphen.is_match(segment) {
                     return Err(format!("Invalid GitLab group/owner segment: '{}'.", segment));
                }
            }
            if !gitlab_name_part_regex.is_match(&repo) || repo.is_empty() || repo.ends_with('.') || no_double_hyphen.is_match(&repo) {
                return Err(format!("Invalid GitLab project name: '{}'.", repo));
            }

            Ok(SourceInfo { owner, repo })
        }
        2 => { // Local File Path
            let path_obj = Path::new(url_str);

            let file_or_dir_name = path_obj.file_name()
                .and_then(|name| name.to_str())
                .map(String::from);

            let repo = match file_or_dir_name {
                Some(name) if !name.is_empty() => name,
                _ => { // If it's a root path like "/" or "C:\" or just an invalid/empty path for filename
                    if path_obj.is_absolute() && path_obj.parent().is_none() { // e.g. / or C:\
                        url_str.to_string() // Use the whole string for root
                    } else {
                        return Err(format!("Invalid local path or could not determine file/directory name: '{}'.", url_str));
                    }
                }
            };
            
            let parent_dir = path_obj.parent()
                .and_then(|p| p.to_str())
                .map(String::from)
                .unwrap_or_else(|| if path_obj.is_absolute() && path_obj.parent().is_none() { "".to_string() } else { ".".to_string() }); // Use "." for relative paths with no explicit parent

            if repo.is_empty() && !(url_str == "/" || (url_str.ends_with(":\\") && url_str.len() == 3) ) { // Check for empty repo unless it's a root
                 return Err(format!("Local path is empty or invalid: '{}'.", url_str));
            }

            Ok(SourceInfo { owner: parent_dir, repo })
        }
        _ => Err("Invalid source type specified. Use 0 for GitHub, 1 for GitLab, 2 for Local File.".to_string()),
    }
}